Improve structure error feedback

RecursivePineapple opened the pull request:
This adds a mechanism for easily adding validation error messages for unique structure requirements. Many multis have specific requirements, but these requirements aren't communicated anywhere and there's no feedback for when the requirements aren't met. This PR allows multis to add error messages easily.

For now, it's only implemented on the cryo freezer and gt++ multis, but if this is accepted other multis can add their own validations easily.

RecursivePineapple commented:
Drafting because I didn't know someone worked on something similar (just saw the merge conflicts). Let me see if this needs to be closed first.

RecursivePineapple commented:
Some examples:
![2025-02-03_13 02 50](https://github.com/user-attachments/assets/9919337b-d2ac-4af0-a623-351e4659a8ad)
![2025-02-03_13 02 45](https://github.com/user-attachments/assets/b59cac8a-14a6-409e-9c88-cdafbcbbea8a)
![2025-02-03_13 02 57](https://github.com/user-attachments/assets/dfd9aed3-a95d-4e29-bc65-2ca97ffa6191)
![2025-02-03_13 03 00](https://github.com/user-attachments/assets/ed1b32b9-f817-4d1e-845a-fd2c8c758ab1)
![2025-02-03_13 03 06](https://github.com/user-attachments/assets/800db32a-0038-4df2-8dbb-9ea034f20a0a)

RecursivePineapple commented:
False alarm, I saw get/set DisplayErrorID and thought it was a similar mechanism but it looks like that's just for generic errors shared by all multis.

YannickMG commented:
With this current implementation it looks like each implementation will need to define exactly how it wants to handle displaying a given error message, and that when said error message is displayed the data needed to explain the error may need to be reacquired if the `StructureError` isn't specific enough.

Letting each implementation decide how to display an error is great, it would make it simple for a given implementation to ditch the text display and use other UI elements to indicate the error instead.

The fact that the data relating to the error needs to be re-acquired is a little bit inconvenient though. It works fine for these simple cases, but I can see this feature being extended later to tell you, for instance, that an invalid block was found at position X, Y & Z during the structure check. In order to display a helpful error message we would effectively need to scan the structure again at render time. It might be helpful if these errors could contain information the information needed to render them.

Now I see you've given this some thought since you labeled `mStructureStatus` a list of "unparameterized" structure errors. What was your reasoning for foregoing parameters and state? I recognize that using EnumSet  requires foregoing parameters but holding information on why a structure check failed shouldn't  be on any kind of hot path that would require it.

YannickMG commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -172,6 +175,9 @@ public abstract class MTEMultiBlockBase extends MetaTileEntity
     private static final int CHECK_INTERVAL = 100; // How often should we check for a new recipe on an idle machine?
     private final int randomTickOffset = (int) (Math.random() * CHECK_INTERVAL + 1);
 
+    /** A list of unparameterized structure errors. */
+    protected EnumSet<StructureError> mStructureStatus = EnumSet.noneOf(StructureError.class);
```
If `mStructureStatus` doesn't tell me whether or not the structure is fine, but is instead a list of unparameterized structure errors, why not name it that so that the explanatory comment becomes redundant?

RecursivePineapple commented:
@YannickMG It's unparameterized mainly because of syncing. The structure check always runs on the server, while the text code always runs on the client. It's trivial to sync an EnumSet since you can easily convert it into a byte array, but trying to sync generic data would be way too complex when all that's needed in the multi is a few FakeSyncWidgets. Maybe you could do something with syncing a json object, but it would still be overkill.

A good example is the casing count. Multis may have one 'main' casing, or they could have several in more advanced structures. If my assumption turns out to be incorrect, and we only need to sync a couple of very simple variables, then we can add another layer or refactor this, but I don't want to make that assertion for now.

YannickMG reviewed (changes requested):
![image](https://github.com/user-attachments/assets/1a51eb59-6841-4ef7-8236-916d027b6903)

Some structure errors are emitted despite not matching a structure requirement.

Perhaps you could move the code to add a specific `StructureError` to be part of the structure definition itself so that there can be no mismatch. Otherwise you'll need to find some other way of ensuring `validateStructure` doesn't need to be overridden by every multi with different requirements, unless of course that's the intent.

YannickMG commented:
> @YannickMG It's unparameterized mainly because of syncing. The structure check always runs on the server, while the text code always runs on the client. It's trivial to sync an EnumSet since you can easily convert it into a byte array, but trying to sync generic data would be way too complex when all that's needed in the multi is a few FakeSyncWidgets. Maybe you could do something with syncing a json object, but it would still be overkill.
> 
> A good example is the casing count. Multis may have one 'main' casing, or they could have several in more advanced structures. If my assumption turns out to be incorrect, and we only need to sync a couple of very simple variables, then we can add another layer or refactor this, but I don't want to make that assertion for now.

That makes perfect sense. EnumSets are just fine in this case and it would be problematic to try to anticipate the needs of future parameterized errors in terms of network packets.

RecursivePineapple commented:
> Otherwise you'll need to find some other way of ensuring validateStructure doesn't need to be overridden by every multi with different requirements, unless of course that's the intent.

Yes, that's the main goal. I figured it would be the simplest way to validate the structure. Most of the structure definition code is provided by structurelib, so I can't add to it easily. I have an in-progress project to help with this, but it wraps structurelib completely (it hasn't been drafted yet but I've used for several projects and it works very well - I want to iron out all the kinks first).

I could add some sort of `List<Predicate<T extends MTEMultiBlockBase>> mStructureValidations` mechanism to all multis, but I thought that would've been overkill when it could easily be replaced by a simple method call.

I hadn't expected steam multis to not require maintenance, I'll fix those. In cases like this, you would either change the superclass to require a boolean field for maintenance hatches, or you'd just not call `super.validateStructure()`.

YannickMG commented:
This code doesn't appear to be foolproof since a machine may simply fail to add encountered hatches to `mMufflerHatches`.
```java
        if (this.getPollutionPerSecond(null) == 0 && !mMufflerHatches.isEmpty()) {
            mStructureErrors.add(StructureError.UNNEEDED_MUFFLER);
        }
```

What's an example of a machine it does work for?

![image](https://github.com/user-attachments/assets/347f5e72-0d94-4ff9-9899-93dc136727a3)

![image](https://github.com/user-attachments/assets/a045ab07-145a-45a0-8a32-815a86bbca64)

RecursivePineapple commented:
I'm not entirely sure what multis use it, it's just what checkHatches used to do so I kept it. Almost every gt++ multi emits pollution so it probably isn't that useful, but I didn't want to remove it without a good reason.

YannickMG reviewed (approved):
Fair enough. This will require additional work on several multis to get the most out of it, but it is a nice capability to have.

miozune reviewed (changes requested):
Regarding syncing parametrized errors, see `IMachineMessage` which is implemented by `CheckRecipeResult` and `ShutDownReason`. Making use of it would allow multis to display more detailed information about error.

Also, as pointed out by Yannick, having duplicated code for checking structure can be easily lead to bugs. StructureLib code is not something we can hook into right now, but at least we can control other checks. I'd imagine something like this would be great:
```java
// Specific to multi
List<StructureError> validateStructure() {
  List<StructureError> errors = super.validateStructure();
  reset();
  if (!checkPiece()) {
    // We can't access StructureLib internals, but at least we know something failed with checks
    errors.add(incorrectStructure);
  }
  if (mCasing < 10) {
    // We know exactly what went wrong
    errors.add(tooFewCasings(mCasing, 10));
  }
  return errors;
}

// MultiblockBase
boolean checkStructure() {
  List<StructureError> errors = validateStructure();
  if (!errors.isEmpty()) {
    mMachine = true;
  }
  return mMachine;
}
```

While current code is great as a starting point, I believe we can use more refinement before going implementing all other multis.

RecursivePineapple commented:
I made the mStructureErrors field private so that we can refactor the system easier if needed and added it to the parameters of validateStructure and getStructureErrors. I also added an NBT parameter so that multis can add arbitrary data (it's synced properly, though it has a bit of latency).

I thought about adding some sort of interface system to hide some implementation details, but it would introduce too much boilerplate for not much gain (a bit easier to refactor and namespacing for the nbt data, which won't be a problem). There aren't any limitations to the current system that I know of (though I will admit it's not the prettiest). I think the simplicity makes up for its lack of prettiness (I'm not a fan of excessive boilerplate).

Something like this:
```
interface IServerErrorStorage {
  void addError(StructureError error);
  
  void addError(StructureError error, NBTTagCompound data);
}

interface IClientErrorStorage {
  boolean hasError(StructureError error);

  NBTTagCompound getErrorData(StructureError error);
}
```

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -173,6 +178,15 @@ public abstract class MTEMultiBlockBase extends MetaTileEntity
     private final int randomTickOffset = (int) (Math.random() * CHECK_INTERVAL + 1);
 
+    /** A list of unparameterized structure errors. */
+    private EnumSet<StructureError> mStructureErrors = EnumSet.noneOf(StructureError.class);
```
We should avoid introducing `m` prefixes for new members, so `structureErrors`

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -173,6 +178,15 @@ public abstract class MTEMultiBlockBase extends MetaTileEntity
+     * Any implementation-defined error data.
+     * Private so that multis have to use the parameters (to make it easier to refactor if needed).
+     */
+    private NBTTagCompound mStructureErrorData = new NBTTagCompound();
```
Same here

miozune replied:
Can this have better naming? Having `structureErrors` and `structureErrorData` looks confusing to me. I'd suggest `structureErrorContext`. And `data` argument in `validateStructure` and `getStructureErrors` can be also renamed to `context` if you like this idea.

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -451,12 +465,60 @@ public boolean checkStructure(boolean aForceReset, IGregTechTileEntity aBaseMeta
         return mMachine;
     }
 
+    protected void validateStructureImpl() {
```
Suffix `Impl` refers to a method which child classes override. The name should be swapped so that child classes will override `validateStructureImpl`.

Also, since the intention is to not let children to override this method, it should be marked as `final`. Keeping `protected` is ok so that it can be called from anywhere.

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -451,12 +465,60 @@ public boolean checkStructure(boolean aForceReset, IGregTechTileEntity aBaseMeta
+     * @param data   Generic data blob generated by {@link #validateStructure}.
+     * @param lines  Add text to this. These lines will be shown in the controller GUI.
+     */
+    @SideOnly(Side.CLIENT)
```
`SideOnly` usages should be restricted where it would cause crash if not present. It can be removed.

miozune replied:
I didn't mean to make it available on server, what I meant is you can just remove this annotation. `@SideOnly` is for stripping the member on runtime, not for marking "Don't use this on server." If the intention is to ensure that it won't get called on server, Java compiler cannot understand this and lets anyone to access it on server side, causing `NoSuchMethodError` on runtime. If any, adding Javadoc saying "Don't use this on server" is enough.

RecursivePineapple replied:
I'll add a javadoc, but this method and its overrides should be stripped out on the server. It's not necessary currently since none of the overrides would cause a ClassNotFoundException when loaded on a server, but it'll be helpful if anyone wants to reference other `SideOnly(Client)` code in it for some reason.

I know that `I18n.format` won't cause class load exceptions, but other methods might and I don't want to assume that this is how it'll continue to work into the future.

eigenraven replied:
I'd go even further than miozune and say that this should return something valid on the server too, there's nothing client-specific needed to find structure errors so I don't think this method should be client-only. E.g. if we want to let opencomputers or commands or unit tests etc. read the errors on the server, it should be possible
The StatCollector translation methods should be basically the same as i18n (the classes just have nonsensical mcp names, they're both localization classes)

RecursivePineapple replied:
@eigenraven Makes sense, I hadn't thought about opencomputers/etc integration (though I can't think of a use case for OC specifically).

I removed the SideOnly and swapped out the `I18n.format`s for StatCollector. Once this PR and my AL refactor PR are merged I'll open another one that should pull the janky system I added in the refactor PR into a proper class (so that text can be localized per player on the server somewhat easily).

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -451,12 +465,60 @@ public boolean checkStructure(boolean aForceReset, IGregTechTileEntity aBaseMeta
+     * Scans {@code errors}, {@code data}, or other fields as needed and emits localized structure error messages.
+     * The {@code errors} and {@code data} params are synced already, but any other fields must be manually synced.
+     * Note that the parameters may not be in sync due to network latency (they are synced separately). You shouldn't
+     * rely on a field being in {@code data} if an error is present.
```
`You shouldn't rely on a field being in {@code data} if an error is present.`
Can this be clarified a bit more? Especially whether "error" refers to `errors` argument or general word is unclear.

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -2713,6 +2775,57 @@ protected String generateCurrentRecipeInfoString() {
+                        }
+                    }
+
+                    return out;
```
I believe it's worth creating dedicated class extending `FakeSyncWidget`, since how `StructureError` gets synced has nothing to do with multiblock.

miozune commented on src/main/java/gtPlusPlus/xmod/gregtech/common/tileentities/machines/multi/processing/MTEIndustrialVacuumFreezer.java:
```diff
@@ -48,30 +56,35 @@ public class MTEIndustrialVacuumFreezer extends GTPPMultiBlockBase<MTEIndustrial
-        super(aName);
-        mFuelStack = FluidUtils.getFluidStack("cryotheum", 1);
-        CASING_TEXTURE_ID = TAE.getIndexFromPage(2, 10);
+    protected MTEIndustrialVacuumFreezer(MTEIndustrialVacuumFreezer prototype) {
```
What's the intention behind the change to the constructor? (besides `mFuelStack` and `CASING_TEXTURE_ID` removal) I mean it works, but also inconsistent with all other MTEs. (Yes MTE system is annoying, I'm just not sure if we want to mix two styles for `newMetaEntity` implementation.)

RecursivePineapple replied:
When I started modding gt5u, I was super confused at why there were two constructors. I think that this pattern fits the intention better, since there's one 'fake'/prototype MTE instance that all 'real' instances are cloned from. Nothing in the code describes this, you just figure it out after a while. I think it'd be a good idea to move all MTEs to this pattern, I just haven't gotten around to suggesting it yet.

miozune replied:
Alright, I don't feel this code bad anymore. I feel registering `Supplier<MTE>` would be better rather than instance of MTE itself, but that's whole another argument.

miozune commented on src/main/java/gregtech/api/metatileentity/implementations/MTEMultiBlockBase.java:
```diff
@@ -451,12 +465,60 @@ public boolean checkStructure(boolean aForceReset, IGregTechTileEntity aBaseMeta
+     * @param lines  Add text to this. These lines will be shown in the controller GUI.
+     */
+    @SideOnly(Side.CLIENT)
+    protected void getStructureErrors(Collection<StructureError> errors, NBTTagCompound data, List<String> lines) {
```
"Get" sounds too generic when it's called solely on client, gathers error strings rather than `StructureError`s, and returns nothing. I'd suggest `gatherStructureErrorMessages` or `addStructureErrorMessages`.

miozune reviewed (commented):
I'm still not entirely convinced by NBT syncing v.s. `IMachineMessage` approach, but it's true that NBT can carry flexible information without requiring distinct class, which would the case for structure errors. I'd like to approve the idea now.

Here are some comments need addressing before merge. Sorry for the late review!

miozune commented on src/main/java/gregtech/api/gui/widgets/StructureErrorSyncer.java:
```diff
@@ -0,0 +1,49 @@
+        super(getter, setter, StructureErrorSyncer::save, StructureErrorSyncer::load);
+    }
+
+    public static void save(PacketBuffer buffer, EnumSet<StructureError> errors) {
```
This should be private

miozune commented on src/main/java/gregtech/api/gui/widgets/StructureErrorSyncer.java:
```diff
@@ -0,0 +1,49 @@
+        buffer.writeBytes(data);
+    }
+
+    public static EnumSet<StructureError> load(PacketBuffer buffer) {
```
Same here

miozune commented on src/main/java/gtPlusPlus/xmod/gregtech/api/metatileentity/implementations/base/GTPPMultiBlockBase.java:
```diff
@@ -490,8 +494,43 @@ public void updateSlots() {
     }
 
+    @SideOnly(Side.CLIENT)
+    protected void localizeStructureErrors(Collection<StructureError> errors, NBTTagCompound context,
```
nit: missing `@Override`

miozune reviewed (commented):
I have noticed that when structure for cryo freezer is incomplete, meaning `checkPiece` fails, all the warnings about maintenance, muffler, cryotheum hatches and the amount of casings pop up, even though they're satisfied. Maybe we can skip `doStructureValidation` if `checkMachine` failed.

miozune reviewed (approved):
Another QoL on the horizon 🚀
